src/psyclone/psyir/nodes/call.py

```diff
@@ -633,69 +637,78 @@ def get_callees(self):
 
     def _check_argument_type_matches(
                 self,
-                call_arg: DataSymbol,
-                routine_arg: DataSymbol,
-            ) -> None:
-        """Return information whether argument types are matching.
-        This also supports 'optional' arguments by using
+                call_arg: DataNode,
+                routine_arg: DataSymbol
+            ):
+        """Checks whether the supplied call and routine arguments are
+        compatible. This also supports 'optional' arguments by using
         partial types.
 
         :param call_arg: One argument of the call
         :param routine_arg: One argument of the routine
 
-        :raises CallMatchingArgumentsNotFound: Raised if no matching argument
-            was found.
+        :raises CallMatchingArgumentsNotFound: if the supplied arguments
+            do not match.
 
         """
-        if isinstance(
-            routine_arg.datatype, UnsupportedFortranType
-        ):
-            # This could be an 'optional' argument.
-            # This has at least a partial data type
-            if (
-                call_arg.datatype
-                != routine_arg.datatype.partial_datatype
-            ):
+        if isinstance(call_arg.datatype, ArrayType) and isinstance(
```

Comment:
I don't understand this part - any arrays are just accepted (e.g. not even comparing the number of dimensions)? If that is a temporary work around, please add a #TODO. If I am missing something, I would appreciate a comment here.

Reply:
No, that wasn't very good. I've updated it now with proper checks.

Reply:
Thanks, now I understand that :)